addons/purchase_stock/models/purchase.py

```diff
@@ -329,6 +331,40 @@ def _compute_forecasted_issue(self):
                 if virtual_available < 0:
                     line.forecasted_issue = True
 
+    @api.onchange('product_packaging_id', 'product_qty', 'product_uom')
+    def _onchange_product_packaging(self):
+        if self.product_packaging_id and self.product_qty:
+            newqty = self.product_packaging_id._check_qty(self.product_qty, self.product_uom, "ceiling")
+            if float_compare(newqty, self.product_qty, precision_rounding=self.product_uom.rounding) != 0:
+                return {
+                    'warning': {
+                        'title': _('Warning'),
+                        'message': _(
+                            "This product is packaged by %(pack_size).2f %(pack_name)s. You should purchase %(quantity).2f %(unit)s.",
+                            pack_size=self.product_packaging_id.qty,
+                            pack_name=self.product_id.uom_id.name,
+                            quantity=newqty,
+                            unit=self.product_uom.name
+                        ),
+                    },
+                }
+
+    @api.onchange('product_packaging_id', 'product_uom', 'product_qty')
+    def _onchange_update_packaging_qty(self):
```

Comment:
try to put it in _onchange_product_packaging.

Check that if warning message is raised, data still update